### PR TITLE
Allow for setting a custom HTTP client

### PR DESCRIPTION
**Pull request by thinkingserious**

### CODE REVIEW

**Comment by thinkingserious** [comment]
@sridharv, @broady, 

Given the hiccups with the CLA and need for a fix, I went ahead and created this pull request. I would appreciate a quick review to ensure this will do what you need.

For reference, this closes:
https://github.com/sendgrid/rest/issues/5
https://github.com/sendgrid/sendgrid-go/issues/63
https://github.com/sendgrid/rest/pull/8
https://github.com/sendgrid/rest/pull/6
https://github.com/sendgrid/sendgrid-go/pull/59

**Comment by sridharv** [comment]
Looks good to me. 

Thanks for porting.

File: rest_test.go [hunk 1]
```diff
@@ -105,3 +108,24 @@ func TestRest(t *testing.T) {
 		t.Errorf("Rest failed to make a valid API request. Returned error: %v", e)
 	}
 }
+
+func TestCustomHTTPClient(t *testing.T) {
+	fakeServer := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		time.Sleep(time.Second * 2)
```
Comment (broady) [inline rest_test.go]: These timeouts can be shorter - try 100ms & 200ms.

**Comment by broady** [comment]
LGTM

File: rest_test.go [hunk 2]
```diff
@@ -105,3 +108,24 @@ func TestRest(t *testing.T) {
...
+	host := fakeServer.URL
+	endpoint := "/test_endpoint"
+	baseURL := host + endpoint
+	method := Get
+	request := Request{
+		Method:  method,
+		BaseURL: baseURL,
+	}
+	customClient := &Client{&http.Client{Timeout: time.Second * 1}}
+	_, e := customClient.API(request)
```
Comment (broady) [inline rest_test.go]: s/e/err/

```
if _, err := customClient.API(request); err == nil {
  t.Error("want err, got nil")
}
```

File: rest_test.go [hunk 3]
```diff
@@ -105,3 +108,24 @@ func TestRest(t *testing.T) {
...
+	if e == nil {
+		t.Error("Timeout did not trigger as expected")
+	}
```
Comment (broady) [inline rest_test.go]: This test doesn't check the type of error returned.

**Comment by thinkingserious** [comment]
Thanks for the feedback! I'll make the suggested changes here and get this deployed and update the dependency in sendgrid-go.

**Comment by thinkingserious** [comment]
@broady @sridharv,

Thank you so much for your support and help on this!

Please send us your mailing address and T-shirt size (dx@sendgrid.com) so we can give you a small token of our appreciation.

**Comment by broady** [comment]
woohoo!

**Pull request merged**